pkg/domain/ru_stats.go

```diff
@@ -0,0 +1,279 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package domain
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/infoschema"
+	"github.com/pingcap/tidb/pkg/kv"
+	"github.com/pingcap/tidb/pkg/meta"
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/util/intest"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	pd "github.com/tikv/pd/client"
+	"go.uber.org/zap"
+)
+
+const (
+	maxRetryCount   int           = 10
+	ruStatsInterval time.Duration = 24 * time.Hour
+	// only keep stats rows for last 3 monthes(92 days at most).
+	ruStatsGCDuration time.Duration = 92 * ruStatsInterval
+	gcBatchSize       int64         = 1000
+)
+
+// RUStatsWriter represents a write to write ru historical data into mysql.request_unit_by_group.
+type RUStatsWriter struct {
+	// make some fields public for unit test.
+	Interval  time.Duration
+	RMClient  pd.ResourceManagerClient
+	InfoCache *infoschema.InfoCache
+	store     kv.Storage
+	sessPool  *sessionPool
+	// current time, cache it here to make unit test easier.
+	StartTime time.Time
+}
+
+// NewRUStatsWriter build a RUStatsWriter from Domain.
+func NewRUStatsWriter(do *Domain) *RUStatsWriter {
+	return &RUStatsWriter{
+		Interval:  ruStatsInterval,
+		RMClient:  do.GetPDClient(),
+		InfoCache: do.infoCache,
+		store:     do.store,
+		sessPool:  do.sysSessionPool,
+	}
+}
+
+func (do *Domain) dailyRequestUnitsWriterLoop() {
```

Comment:
```suggestion
func (do *Domain) RequestUnitsWriterLoop() {
```
it relies on `ruStatsInterval`